### PR TITLE
Add Patch resources to all SDKs

### PR DESCRIPTION
**Pull request by lblackstone**
### Proposed changes
Add Patch resources for all Kubernetes resource types that can be used to patch existing cluster resources rather than creating a new resource under Pulumi management. See https://github.com/pulumi/pulumi-kubernetes/issues/2011 for additional details on the background and design of this feature.

### Related issues (optional)
Fix https://github.com/pulumi/pulumi-kubernetes/issues/2015

### CODE REVIEW

**Comment by lblackstone** [comment]
@viveklak It looks like tests may be failing due to parallel resource creation across SDKs. I'll take another look tomorrow, but they're passing for me locally.

File: provider/pkg/await/await.go [hunk 1]
```diff
@@ -181,8 +181,9 @@ func Creation(c CreateConfig) (*unstructured.Unstructured, error) {
 			if c.ServerSideApply {
 				force := metadata.IsAnnotationTrue(c.Inputs, metadata.AnnotationPatchForce)
 				options := metav1.PatchOptions{
-					FieldManager: c.FieldManager,
-					Force:        &force,
+					FieldManager:    c.FieldManager,
+					Force:           &force,
+					FieldValidation: metav1.FieldValidationIgnore,
```
Comment (viveklak) [inline provider/pkg/await/await.go:187]: Do we want this to be warn instead?

File: tests/sdk/nodejs/server-side-apply/step2/index.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+// Patch the CronTab CustomResource to add a label.
+export const crPatched = new k8s.apiextensions.CustomResourcePatch("label-cr",
+    {
+        apiVersion: "stable.example.com/v1",
+        kind: "CronTab",
+        metadata: {
+            labels: {
+                foo: "foo",
+            },
+            name: "my-new-cron-object",
```
Comment (viveklak) [inline tests/sdk/nodejs/server-side-apply/step2/index.ts:103]: Just wanted to make sure we are not letting patch resources by autonamed in the provider? I did notice that we are not enforcing its requirement in the SDK generation (which maybe we ought to do?)

File: provider/pkg/gen/schema.go [hunk 3]
```diff
@@ -450,6 +531,17 @@ func genPropertySpec(p Property, resourceGV string, resourceKind string) pschema
 	err := json.Unmarshal([]byte(p.SchemaType()), &typ)
 	contract.Assert(err == nil)
 
+	if strings.HasSuffix(resourceKind, "Patch") {
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/gen/schema.go:534]: Yeah, there's nothing currently, and I think we could handle that case if it did arise in a later release. I don't think it's particularly likely.

File: tests/sdk/nodejs/server-side-apply/step2/index.ts (hunk 2 above)
Comment (lblackstone) [reply tests/sdk/nodejs/server-side-apply/step2/index.ts:103]: Good point. The name should be required.

File: provider/pkg/gen/schema.go [hunk 4]
```diff
@@ -294,6 +298,44 @@ func PulumiSchema(swagger map[string]interface{}) pschema.PackageSpec {
...
+					patchSpec = pschema.ObjectTypeSpec{
+						Description: kind.Comment() + kind.PulumiComment(),
+						Type:        "object",
+						Properties:  map[string]pschema.PropertySpec{},
+						Language:    map[string]pschema.RawMessage{},
+					}
+
+					var propNames []string
+					for _, p := range kind.Properties() {
+						patchSpec.Properties[p.name] = genPropertySpec(p, kind.apiVersion, kind.kind+"Patch")
```
Comment (viveklak) [inline provider/pkg/gen/schema.go:314]: Somewhat subtle that no required properties are being set (by design for patch resources). Might be worth calling out with a comment. We might want to add name and/or namespace (for namespaced objects back though)?

File: provider/pkg/provider/provider.go [hunk 5]
```diff
@@ -2683,12 +2707,13 @@ func (k *kubeProvider) tryServerSidePatch(
 			return nil, nil, false, err
 		}
 	}
-	if err != nil {
-		return nil, nil, false, err
+	if err.Error() == "name is required" {
```
Comment (viveklak) [inline provider/pkg/provider/provider.go:2710]: Nit - consider returning a strongly typed error for this instead?

**Review by viveklak — APPROVED** [review]
Directionally LGTM! We might move addressing some of the suggestions to a subsequent PR for review ergonomics.

File: provider/pkg/gen/dotnet-templates/apiextensions/CustomResourcePatch.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+// limitations under the License.
+
+// *** WARNING: this file was generated by pulumigen. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+using System.Collections.Immutable;
+using Pulumi.Kubernetes.Types.Inputs.Meta.V1;
+using Pulumi.Kubernetes.Types.Outputs.Meta.V1;
+
+namespace Pulumi.Kubernetes.ApiExtensions
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/gen/dotnet-templates/apiextensions/CustomResourcePatch.cs:22]: I just copy/pasted the implementation from CustomResource and renamed some relevant fields. I think it's ok for an initial pass here.

**Comment by lblackstone** [comment]
> Directionally LGTM! We might move addressing some of the suggestions to a subsequent PR for review ergonomics.

Sounds good! I will make another pass through these feedback items once this merges.

**Pull request merged**